br/pkg/lightning/restore/data_sample_check.go

```diff
@@ -0,0 +1,284 @@
+					for i, col := range fileInfo.TableInfo.Columns {
+						columnNames[i] = col.Name.L
+					}
+				} else {
+					columnCount = len(columnNames)
+				}
+			}
+			rowsChecked++
+			if len(row.Row) != columnCount {
+				columnCountMismatchRows++
+				d.logCheckFailedRow(row.Row, fileMeta, lastOffset,
+					fmt.Sprintf("column count mismatch, expect %d, got %d", columnCount, len(row.Row)))
+			}
+
+			for colIdx, col := range row.Row {
+				if colIdx >= columnCount {
+					break
+				}
+
+				columnInfo := ddlColumnMap[columnNames[colIdx]]
+				if columnInfo == nil {
+					// we don't check this kind of error here
+					continue
+				}
+				colType := columnInfo.FieldType
+				// we only check non-binary string types here
+				if !types.IsNonBinaryStr(&colType) {
+					continue
+				}
+				colStr := col.GetString()
+				if !utf8.ValidString(colStr) || strings.Contains(colStr, config.DefaultCSVDataInvalidCharReplace) {
+					invalidCharRows++
+					d.logCheckFailedRow(row.Row, fileMeta, lastOffset, "contains invalid char")
+				}
+			}
+			fileParser.RecycleRow(row)
+			// lastOffset is the start offset of current row
+			lastOffset, _ = fileParser.Pos()
+		}
+
+		fileParser.Close()
+
+		d.totalRows.Add(rowsChecked)
+		d.totalColumnCountMismatchRows.Add(columnCountMismatchRows)
+		d.totalInvalidCharRows.Add(invalidCharRows)
+	}
+}
+
+func (d *dataSampleCheck) doCheck(ctx context.Context) error {
+	rc := d.controller
+
+	targetFiles, err := d.getSampledDataFiles()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	concurrency := utils.MinInt(runtime.NumCPU(), 8)
+	fileChan := make(chan *sampledDataFileInfo, concurrency)
+	errChan := make(chan error, concurrency)
+	newCtx, cancelFunc := context.WithCancel(ctx)
+
+	for i := 0; i < concurrency; i++ {
+		d.wg.Add(1)
+		go d.checkRoutine(newCtx, fileChan, errChan)
+	}
+
+	for _, fileInfo := range targetFiles {
+		fileChan <- fileInfo
+	}
+	close(fileChan)
+
+	// we need to exit if any error occurs in check routine
+	// each routine should return a result in the error channel, return nil if there's no error.
+	for i := 0; i < concurrency; i++ {
+		if err = <-errChan; err != nil {
+			cancelFunc()
+			d.wg.Wait()
+
+			log.L().Error("meet error when checking", zap.Error(err))
+			fmt.Printf("failed to do data sample check: %s\n", err.Error())
+			return errors.Trace(err)
+		}
+	}
+
+	d.wg.Wait()
+	cancelFunc()
+
+	passed := d.totalColumnCountMismatchRows.Load() == 0 && d.totalInvalidCharRows.Load() == 0
+	msg := fmt.Sprintf("Total sample of %d rows of data checked, %d errors found.",
+		d.totalRows.Load(), d.totalColumnCountMismatchRows.Load()+d.totalInvalidCharRows.Load())
+	rc.checkTemplate.Collect(Critical, passed, msg)
+
+	if rc.tidbGlue.OwnsSQLExecutor() {
+		fmt.Println(rc.checkTemplate.Output())
+
+		if rc.checkTemplate.Success() {
+			fmt.Println("All checks have been passed, but there may still be other types of errors that can only be found during the actual insertion of data.")
```

Comment:
This message is misleading if `rc.checkTemplate.Output` contains failing items.

Reply:
success is checking critical failed count, if there're warn failed count, should we take it as `All checks have been passed` @sunzhaoyang

Reply:
warning should be categories to success and no need to resulting in failed result.

Reply:
OK. Then, I think we should also take the result of data sample check into consideration. If data sample check failed, I think the check summary result should be fail instead of pass.

BTW, is it possible to also put the data sample check result into the precheck result table? Use a separate log line is not very user friendly.

Reply:
> OK. Then, I think we should also take the result of data sample check into consideration. If data sample check failed, I think the check summary result should be fail instead of pass.
> 
> BTW, is it possible to also put the data sample check result into the precheck result table? Use a separate log line is not very user friendly.

it is now. In check-only, sample-check and precheck are using the same `checkTemplate`, and precheck doesn't print anything, deferred to sample-check.